crates/astria-sequencer-relayer/tests/blackbox/helpers/test_sequencer_relayer.rs

```diff
@@ -0,0 +1,742 @@
+        let (value, _status_code) = self
+            .wait_until_relayer_api_matches(
+                zpage.to_string().as_str(),
+                within_ms,
+                context,
+                predicate,
+            )
+            .await;
+        let zpage: ZPage = serde_json::from_value(value).expect("should parse {zpage}");
+        zpage.status
+    }
+
+    /// Polls the given endpoint of the sequencer-relayer until it responds with data satisfying
+    /// `predicate`.
+    ///
+    /// The predicate is passed the JSON value and status code from the response, and if it returns
+    /// `true`, the function returns.
+    ///
+    /// # Panics
+    ///
+    /// Panics if any individual request exceeds 100 milliseconds, cannot be parsed, or if the
+    /// predicate isn't satisfied within the number of milliseconds specified.
+    async fn wait_until_relayer_api_matches<P>(
+        &self,
+        api_endpoint: &str,
+        within_ms: u64,
+        context: &str,
+        mut predicate: P,
+    ) -> (serde_json::Value, StatusCode)
+    where
+        P: FnMut(serde_json::Value, StatusCode) -> bool,
+    {
+        let getter = async {
+            loop {
+                let (value, status_code) = self.get_from_relayer_api(api_endpoint, context).await;
+                if predicate(value.clone(), status_code) {
+                    return (value, status_code);
+                }
+                yield_now().await;
+            }
+        };
+        self.timeout_ms(within_ms, context, getter).await
+    }
+
+    /// Gets the JSON body reported via the given endpoint of the sequencer-relayer.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the response cannot be retrieved within 100 milliseconds.
+    async fn get_from_relayer_api(
+        &self,
+        api_endpoint: &str,
+        context: &str,
+    ) -> (serde_json::Value, StatusCode) {
+        let url = format!("http://{}/{api_endpoint}", self.api_address);
+        let getter = async {
+            reqwest::get(&url)
+                .await
+                .unwrap_or_else(|error| panic!("should get response from `{url}`: {error}"))
+        };
+
+        let new_context = format!("{context}: get from `{url}`");
+        let response = self.timeout_ms(100, &new_context, getter).await;
+
+        let status_code = response.status();
+        let value = response
+            .json::<serde_json::Value>()
+            .await
+            .unwrap_or_else(|error| {
+                panic!("{context}: failed to parse response from `{url}` as JSON: {error}")
+            });
+        (value, status_code)
+    }
+
+    /// Executes `future` within the specified duration, returning its result.
+    ///
+    /// If execution takes more than 80% of the allowed time, an error is logged before returning.
+    ///
+    /// # Panics
+    ///
+    /// Panics if execution takes longer than the specified duration.
+    pub async fn timeout_ms<F: Future>(
```

Comment:
So the idea behind this is to grab the `/status` in case of timeouts and report that?

Reply:
Exactly.  I found it pretty useful when writing/debugging the tests.  I also wanted to have a way of seeing if I was close to bumping into any of the timeouts, hence the error message if we consume >80% of the timeout duration.